### PR TITLE
chore: types, types, types

### PR DESCRIPTION
**Pull request by JustinBeckwith**
Not complete at all, but a good start towards typing

### CODE REVIEW

**Comment by JustinBeckwith** [comment]
hah, oh yeah - I need to wait until https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24650 lands :)

**Comment by JustinBeckwith** [comment]
This is ready for a looksy now 👍

**Comment by JustinBeckwith** [comment]
Sweet.  @callmehiphop @alexander-fenster mind taking a gander?

**Comment by JustinBeckwith** [comment]
@googleapis/node-team PTAL :)

File: src/operation.ts [hunk 1]
```diff
@@ -26,7 +26,21 @@ const modelo = require('modelo');
  * @type {module:common/serviceObject}
  * @private
  */
-const ServiceObject = require('./service-object.js');
+const ServiceObject = require('./service-object');
+
+export interface OperationConfig {
```
Comment (kjin) [inline src/operation.ts]: Can you add a TODO to make these fields more descriptive/possibly optional?

File: src/logger.ts [hunk 2]
```diff
@@ -27,35 +27,50 @@ const logDriver = require('log-driver');
...
- * @param {Array.<string>=} options.levels - The list of levels to use. (Default:
- *     logger.LEVELS)
+ * @param {string=|object=} options - Configuration object. If a string, it is treated as `options.level`.
+ * @param {string=} options.level - The minimum log level that will print to the console. (Default: `error`)
+ * @param {Array.<string>=} options.levels - The list of levels to use. (Default: logger.LEVELS)
  * @param {string=} options.tag - A tag to use in log messages.
  */
-function logger(options) {
+function logger(options?: LoggerOptions|string) {
+  let opts: LoggerOptions;
```
Comment (kjin) [inline src/logger.ts]: Does
```ts
if (is.string(options)) {
  options = { level: options as string };
}
```
not work? We should prefer not to have `options` and `opts` in the same scope.

File: src/operation.ts [hunk 3]
```diff
@@ -141,8 +155,8 @@ Operation.prototype.listenForEvents_ = function() {
  *
  * @param {function} callback
  */
-Operation.prototype.poll_ = function(callback) {
-  this.getMetadata((err, resp) => {
+Operation.prototype.poll_ = function(callback: (err?: Error|null, resp?: any) => void) {
```
Comment (kjin) [inline src/operation.ts]: Nit: We should standardize the callback type pattern here and elsewhere. I recommend `(err: Error|null, resp?: any) => void` (though parameterizing the function on the type of resp would be better).

File: src/service-object.ts [hunk 4]
```diff
@@ -149,52 +197,60 @@ ServiceObject.prototype.delete = function(callback) {
 
   // The `request` method may have been overridden to hold any special behavior.
   // Ensure we call the original `request` method.
-  ServiceObject.prototype.request.call(this, reqOpts, (err, resp) => {
+  ServiceObject.prototype.request.call(this, reqOpts, (err: Error|null, resp: r.Response) => {
```
Comment (kjin) [inline src/service-object.ts]: Same comment as above regarding callbacks.

File: src/logger.ts (hunk 2 above)
Comment (JustinBeckwith) [reply src/logger.ts]: Fixed this with proper TypeScript method overloading.  Good call.

File: src/operation.ts (hunk 3 above)
Comment (JustinBeckwith) [reply src/operation.ts]: I hear you.  Step 1 is to kind of identify what the various return types are across the module.  It's a little scattered and unclear.  After we're confident we have all the right types in place, I'm 100% for distilling it down to a rational set of types.

**Pull request merged**